ocaml/libs/http-lib/http.ml

```diff
@@ -449,73 +434,120 @@ let read_http_response_header buf fd =
-    ; (* None means '*' *)
-      subty: string option
+      ty: (string * string option) option
     ; (* None means '*' *)
       q: int
           (* range 0 - 1000 *)
           (* We won't parse the more advanced stuff *)
   }
 
-  let string_of_t x =
-    Printf.sprintf "%s/%s;q=%.3f"
-      (Option.value ~default:"*" x.ty)
-      (Option.value ~default:"*" x.subty)
-      (float_of_int x.q /. 1000.)
-
-  let matches (ty, subty) = function
-    | {ty= Some ty'; subty= Some subty'; _} ->
-        ty' = ty && subty' = subty
-    | {ty= Some ty'; subty= None; _} ->
-        ty' = ty
-    | {ty= None; subty= Some _; _} ->
-        assert false
-    | {ty= None; subty= None; _} ->
+  let equal_ty a b =
+    match (a, b) with
+    | Some _, None
+    | Some (_, Some _), Some (_, None)
+    | None, Some _
+    | Some (_, None), Some (_, Some _) ->
+        false
+    | None, None ->
+        true
+    | Some (a, None), Some (b, None) ->
+        String.equal a b
+    | Some (a, Some a_s), Some (b, Some b_s) ->
+        if not (String.equal a b) then false else String.equal a_s b_s
+
+  let equal {ty= a_ty; q= a_q} {ty= b_ty; q= b_q} =
+    if not (Int.equal a_q b_q) then
+      false
+    else
+      equal_ty a_ty b_ty
+
+  let to_string x =
+    let ty, subty =
+      match x.ty with
+      | None ->
+          ("*", "*")
+      | Some (ty, subty) ->
+          (ty, Option.value ~default:"*" subty)
+    in
+    Printf.sprintf "%s/%s;q=%.3f" ty subty (float_of_int x.q /. 1000.)
+
+  let ( // ) = Filename.concat
+
+  let matches ty = function
+    | {ty= Some (ty', Some subty'); _} ->
+        String.equal (ty' // subty') ty
+    | {ty= Some (ty', None); _} -> (
+      match String.split_on_char '/' ty with
+      | [] ->
+          false
+      | ty :: _ ->
+          String.equal ty' ty
+    )
+    | {ty= None; _} ->
         true
 
-  (* compare [a] and [b] where both match some media type *)
-  let compare (a : t) (b : t) =
-    let c = compare a.q b.q in
+  let compare_user_preference (a : t) (b : t) =
+    let c = Stdlib.compare a.q b.q in
     if c <> 0 then
       -c (* q factor (user-preference) overrides all else *)
     else
       match (a.ty, b.ty) with
-      | Some _, None ->
-          1
-      | None, Some _ ->
+      | Some _, None | Some (_, Some _), Some (_, None) ->
           -1
-      | _, _ -> (
-        match (a.subty, b.subty) with
-        | Some _, None ->
-            1
-        | None, Some _ ->
-            -1
-        | _, _ ->
-            0
-      )
+      | None, Some _ | Some (_, None), Some (_, Some _) ->
+          1
+      | None, None
+      | Some (_, None), Some (_, None)
+      | Some (_, Some _), Some (_, Some _) ->
+          0
+
+  (** [pick_preferred ~allowed accepted] filters [accepted] removing typed that
+      do not match [accepted] and orders them by client weighting *)
+  let pick_preferred ~allowed types =
+    (* Because compare_user_preference is insesible to the actual value of the
```

Comment:
Would `List.stable_sort` help?

Reply:
The main issue is that the list of matches for each mimetype need to be merged in a way that deduplicates while maintaining the original sorting order.

Using stable sort doesn't help you in some cases when the merging happens because the stable sort happens on the merged list, not the original one, in other words, the algorithm is still sensible to the order of the list in `~allowed`

For example:
``` 
let allowed = ["text/xml"; "application/json"]
let types = "*/*;q=0.8,application/json,text/xml,text/html"
```
Because `text/xml` in the first item of `allowed`, its filtered list for compatible types filters out `application/json`, leaving the first item to be `text/xml`. This means that the merged list begins with `text/xml`, which stable sort will maintain: 
```
   Expected: `[application/json;q=1.000; text/xml;q=1.000; */*;q=0.800]'
   Received: `[text/xml;q=1.000; application/json;q=1.000; */*;q=0.800]'
```

Reply:
Another alternative would be to accumulate items that do not match the allowed mimetypes, on the on merge keep the ones that are contained in all the lists, and finally remove those from the end list. But this seems more complex than keeping track of the indices of items that must be kept

Reply:
What are the semantics that we need? Set operations on ordered lists? I'm not even sure what that means but I suspect that we would like some algebra over lists where order is meaningful. Would it make sense to write a module for that?

Reply:
I've changed the algorithm by sorting the accept list by user preference then consuming elements from one priority at a time to select the content types that match them, returning the first non-empty result, or an empty one if all the list has been consumed.